Tetrahedral_remeshing/include/CGAL/Tetrahedral_remeshing/Adaptive_remeshing_sizing_field.h

```diff
@@ -0,0 +1,339 @@
+ * A sizing field, model of `RemeshingSizingField_3`
+ */
+template <typename Tr>
+class Adaptive_remeshing_sizing_field
+  : public Sizing_field<typename Tr::Geom_traits>
+{
+  // Types
+  typedef typename Tr::Geom_traits              GT;
+  typedef typename Tr::Geom_traits::Point_3     Bare_point;
+  typedef typename Tr::Point                    Tr_point;
+  typedef typename GT::FT                       FT;
+
+  typedef typename Tr::Vertex_handle            Vertex_handle;
+  typedef typename Tr::Cell_handle              Cell_handle;
+
+  struct Point_with_info
+  {
+    Bare_point p;
+    FT size;
+    int dimension;
+  };
+
+private:
+  struct Point_property_map
+  {
+    using Self = Point_property_map;
+    using value_type = Bare_point;
+    using reference = value_type; //TODO : why can't that be value_type& ?
+    using key_type = Point_with_info;
+    using category = boost::readable_property_map_tag;
+
+    const value_type operator[](const key_type& pwi) const { return pwi.p; }
+    friend const value_type get(const Self&, const key_type& pwi) { return pwi.p; }
+  };
+
+private:
+  using Kd_traits = CGAL::Search_traits_adapter<Point_with_info,
+                                                Point_property_map,
+                                                CGAL::Search_traits_3<GT> >;
+  using Neighbor_search = CGAL::Orthogonal_k_neighbor_search<Kd_traits>;
+  using Kd_tree = typename Neighbor_search::Tree;
+  using Distance = typename Neighbor_search::Distance;
+  using Splitter = typename Neighbor_search::Splitter;
+
+protected:
+  template<typename ECMap, typename FCMap, typename CellSelector>
+  Adaptive_remeshing_sizing_field(const Tr& tr,
+                                  const ECMap& ecmap,
+                                  const FCMap& fcmap,
+                                  const CellSelector& cell_selector)
+    : m_kd_tree(points_with_info(tr, ecmap, fcmap, cell_selector),
+                Splitter(),
+                Kd_traits(Point_property_map()))
+  {
+    m_kd_tree.build();
+  }
+
+private:
+
+  template<typename ECMap, typename FCMap, typename CellSelector>
+  std::vector<Point_with_info> points_with_info(const Tr& tr,
+                                                const ECMap& ecmap,
+                                                const FCMap& fcmap,
+                                                const CellSelector& cell_selector) const
+  {
+    auto cp = tr.geom_traits().construct_point_3_object();
+    std::vector<Point_with_info> points;
+    for (const Vertex_handle v : tr.finite_vertex_handles())
+    {
+      points.push_back(
+        Point_with_info{ cp(tr.point(v)),
+                         average_edge_length_around(v, tr, ecmap, fcmap, cell_selector),
+                         v->in_dimension() });
+    }
+    return points;
+  }
+
+public:
+  /**
+  * Returns size at point `p`
+  */
+  template <typename Index>
+  FT operator()(const Bare_point& p, const int& dim, const Index& /* i */) const
```

Comment:
please double check that it is expected that `i` is not used

Reply:
yes it's fine